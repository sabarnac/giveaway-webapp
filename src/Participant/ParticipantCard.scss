.participant-card {
  display: flex;
  flex-direction: column;
  justify-content: center;
  .participant-card__avatar {
    text-align: center;
    height: 5rem;
    img {
      height: 5rem;
    }
  }
  .participant-card__name {
    text-align: center;
    font-size: 3rem;
    word-break: break-all;
  }
  padding: 2rem 2rem 1rem 2rem;
  border: 1px solid #aaa;
  width: 45%;
}
