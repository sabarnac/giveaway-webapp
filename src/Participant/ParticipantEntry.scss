.participant-entry {
  display: flex;
  justify-content: space-evenly;
  align-items: center;
  .participant-entry__avatar {
    text-align: center;
    height: 5rem;
    img {
      height: 5rem;
    }
    width: 50%;
  }
  .participant-entry__name {
    text-align: center;
    font-size: 3rem;
    width: 50%;
    word-break: break-all;
  }
  padding: 2rem;
  border: 1px solid #aaa;
  width: 45%;
}
