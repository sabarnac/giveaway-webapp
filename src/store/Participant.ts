import { observable, computed } from "mobx";

/**
 * Class containing the details of a participants' avatar.
 */
export class Avatar {
  @observable private _url: string;
  @observable private _altText: string;

  public constructor(url: string, altText: string) {
    this._url = url;
    this._altText = altText;
  }

  /**
   * Gets the avatar URL.
   * @return {string} The avatar URL.
   */
  public get url(): string {
    return this._url;
  }

  /**
   * Gets the avatar image alternate text.
   * @return {string} The avatar image alternate text.
   */
  public get altText(): string {
    return this._altText;
  }

  private isSameUrl = (url1: string, url2: string): boolean => url1 === url2;

  private isSameAltText = (altText1: string, altText2: string): boolean =>
    altText1 === altText2;

  /**
   * Determines whether another avatar is identical to the current one.
   * @param  {Avatar} otherAvatar The other avatar to compare against.
   * @returns {boolean} Whether the other avatar is equal to the current one.
   */
  public equals = (otherAvatar: Avatar): boolean =>
    this.isSameUrl(this._url, otherAvatar._url) &&
    this.isSameAltText(this._altText, otherAvatar._altText);
}

/**
 * Class containing the details of a participant.
 */
export default class Participant {
  @observable private _name: string;
  @observable private _avatar?: Avatar;

  public constructor(name: string, avatar?: Avatar) {
    this._name = name;
    this._avatar = avatar;
  }

  /**
   * Gets the participant name.
   * @return {string} The participant name.
   */
  public get name(): string {
    return this._name;
  }

  /**
   * Gets the participant avatar, if it exists.
   * @return {string | null} The participant avatar, or null if there isn't one.
   */
  public get avatar(): Avatar | null {
    return this._avatar || null;
  }

  /**
   * Gets whether the participant has an avatar or not.
   * @returns {boolean} Whether the participant has an avatar or not.
   */
  @computed public get hasAvatar(): boolean {
    return !!this._avatar;
  }

  /**
   * Gets the capitalized name of the participant.
   * @returns {string} The capitalized name.
   */
  @computed public get properName(): string {
    return this._name.replace(/\w\S*/g, (subText: string) => {
      return subText.charAt(0).toUpperCase() + subText.substr(1).toLowerCase();
    });
  }

  private isSameName = (name1: string, name2: string): boolean =>
    name1 === name2;

  private isSameAvatar = (avatar1?: Avatar, avatar2?: Avatar): boolean =>
    avatar1 && avatar2
      ? avatar1.equals(avatar2)
      : !avatar1 && !avatar2
      ? true
      : false;

  /**
   * Determines whether another participant is identical to the current one.
   * @param  {Participant} otherParticipant The other participant to compare against.
   * @returns {boolean} Whether the other participant is equal to the current one.
   */
  public equals = (otherParticipant: Participant): boolean =>
    this.isSameName(this._name, otherParticipant._name) &&
    this.isSameAvatar(this._avatar, otherParticipant._avatar);
}
