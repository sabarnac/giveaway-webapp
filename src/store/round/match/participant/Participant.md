Usage:

```tsx
import Avatar from "./store/round/match/participant/Avatar";

const participant: Participant = new Participant(
  "FOO BAR" /*, {url:"url",altText:"alt-text"} */
);
const participantName: string = participant.name; // "FOO BAR"
const participantProperName: string = participant.properName; // "Foo Bar"
const participantAvatar: Avatar = participant.avatar; // Avatar("url","alt-text")
```
