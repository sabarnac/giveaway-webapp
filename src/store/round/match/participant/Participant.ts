import { observable, computed } from "mobx";
import { createAvatarImage } from "../../../../util";
import Avatar from "./Avatar";
const inflect = require("i")();

interface AvatarJson {
  url: string;
  altText: string;
}

/**
 * Class containing the details of a participant.
 */
export default class Participant {
  /** The ID of the participant. */
  @observable private _id: string;
  /** The name of the participant. */
  @observable private _name: string;
  /** The avatar of the participant. */
  @observable private _avatar: Avatar;
  /** The weight of the participant. */
  @observable private _weight: number;

  /** A counter for generating a unique ID for the match. */
  private static counter: number = 1;

  /**
   * Gets the avatar of the participant if present, or creates a random one,
   * @param  {string} name The name of the participant.
   * @param  {AvatarJson} avatar The avatar details of the participant, if available.
   * @return {Avatar} The generated avatar of the participant.
   */
  private _getOrCreateAvatar = (name: string, avatar?: AvatarJson): Avatar => {
    return avatar
      ? new Avatar(avatar.url, avatar.altText)
      : new Avatar(createAvatarImage(name), name);
  };

  public constructor(name: string, avatar?: AvatarJson, weight: number = 1) {
    this._id = `${Participant.counter++}`;
    this._name = name;
    this._avatar = this._getOrCreateAvatar(name, avatar);
    this._weight = weight;
  }

  /**
   * Get the participant ID.
   * @return {string} The unique ID of the participant.
   */
  @computed public get id(): string {
    return this._id;
  }

  /**
   * Gets the participant name.
   * @return {string} The participant name.
   */
  @computed public get name(): string {
    return this._name;
  }

  /**
   * Gets the capitalized name of the participant.
   * @return {string} The capitalized name.
   */
  @computed public get properName(): string {
    return inflect.titleize(this._name);
  }

  /**
   * Gets the participant avatar, if it exists.
   * @return {string} The participant avatar.
   */
  @computed public get avatar(): Avatar {
    return this._avatar;
  }

  /**
   * Gets the participant weight.
   * @return {number} The participant weight.
   */
  @computed public get weight(): number {
    return this._weight;
  }

  /**
   * Returns whether the two names are equal.
   * @param {string} name1 The first name to compare.
   * @param {string} name2 The second name to compare.
   * @return {boolean} Whether the two names are equal.
   */
  private _isSameName = (name1: string, name2: string): boolean =>
    name1 === name2;

  /**
   * Returns whether the two avatars are equal.
   * @param {Avatar} avatar1 The first avatar to compare.
   * @param {Avatar} avatar2 The second avatar to compare.
   * @return {boolean} Whether the two avatars are equal.
   */
  private _isSameAvatar = (avatar1: Avatar, avatar2: Avatar): boolean =>
    avatar1.equals(avatar2);

  /**
   * Determines whether another participant is identical to the current one.
   * @param  {Participant} otherParticipant The other participant to compare against.
   * @return {boolean} Whether the other participant is equal to the current one.
   */
  public equals = (otherParticipant: Participant): boolean =>
    this._isSameName(this._name, otherParticipant._name) &&
    this._isSameAvatar(this._avatar, otherParticipant._avatar);
}
