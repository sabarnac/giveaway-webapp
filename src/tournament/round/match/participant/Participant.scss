.participant {
    background: #ffffff;
    &--invert {
        background: #000000;
        color: #ffffff;
    }
}