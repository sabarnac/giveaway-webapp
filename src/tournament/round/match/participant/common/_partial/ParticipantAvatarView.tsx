import React from "react";
import { observer } from "mobx-react";
import classNames from "classnames";
import Participant from "../../../../../../store/round/match/participant/Participant";

/**
 * Properties of the participant card React component.
 */
interface ParticipantAvatarViewProps {
  /** CSS class */
  className: string;
  /** The participant details. */
  participant: Participant;
  /** Whether to invert the color of the participant view or not. */
  invert?: boolean;
}

/**
 * Function that returns a React component for the participant view.
 */
export default observer(
  (props: ParticipantAvatarViewProps): JSX.Element => (
    <div className={classNames(`${props.className}__avatar`)}>
      <img
        src={props.participant.avatar.url}
        alt={props.participant.avatar.altText}
      />
    </div>
  ),
);
