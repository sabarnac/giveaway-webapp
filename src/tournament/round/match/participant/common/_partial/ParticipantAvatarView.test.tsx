import React from "react";
import ParticipantAvatarView from "./ParticipantAvatarView.react";
import { create, ReactTestRenderer } from "react-test-renderer";
import { createDummyParticipant } from "../../../../../../util/test/index";
import Participant from "../../../../../../store/round/match/participant/Participant";

describe("ParticipantAvatarView Component.", () => {
  it("Matches snapshot.", () => {
    const participant: Participant = createDummyParticipant();
    const component: ReactTestRenderer = create(
      <ParticipantAvatarView className="foobar" participant={participant} />,
    );

    expect(component.toJSON()).toMatchSnapshot();
  });
});
