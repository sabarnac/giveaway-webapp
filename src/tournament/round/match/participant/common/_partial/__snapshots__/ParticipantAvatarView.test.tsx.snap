// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ParticipantAvatarView Component. Matches snapshot. 1`] = `
<div
  className="foobar__avatar"
>
  <img
    alt="fozbaz-1"
    src="https://api.adorable.io/avatars/480/fozbaz-1"
  />
</div>
`;
